File: day01/day01.rb
def part_one
  seen = {}
  numbers = File.new('input.txt').readlines.map(&:to_i)
  solution = find_two(numbers, 2020)
  puts(solution[0] * solution[1])
end

def find_two(l, sum)
  seen = {}
  l.each do |n|
    needed = sum - n
    if seen.key? needed
      return [needed, n]
    else
      seen[n] = 1
    end
  end
  return []
end

def part_two
  seen = {}
  lines = File.new('input.txt').readlines.map(&:to_i)
  sorted = lines.sort
  0.upto(sorted.length - 2).each do |i|
    needed = 2020 - (sorted[i] + sorted[i + 1])
    sorted.each do |n|
      if n == needed
        puts(sorted[i] * sorted[i + 1] * needed)
      end
    end
  end
end

part_two

File: python/day01/day01.py
def read_file(filename):
    """Read file, converting each line into an integer"""
    fp = open(filename, 'r')
    lines = fp.readlines()
    fp.close()
    return list(map(lambda l: int(l.strip()), lines))

def find_two(numbers, sumval):
    """Find two numbers in numbers that sum to sumval."""
    seen = {}
    for number in numbers:
        needed = sumval - number
        if needed in seen:
            return (number, needed)
        else:
            seen[number] = True
    return ()

def find_three(numbers, sumval):
    """Find three numbers in numbers that sum to sumval."""
    seen = {}
    for number in numbers:
        needed = sumval - number
        seen[number] = needed
        attempt = find_two(numbers, needed)
        if len(attempt) == 2:
            return (attempt[0], attempt[1], number)
    return ()

input = read_file('input.txt')

# Part one
result = find_two(input, 2020)
if len(result) == 2:
    print(result[0] * result[1])

# Part two
result = find_three(input, 2020)
if len(result) == 3:
    print(result[0] * result[1] * result[2])